### ISSUE TITLE
About newer versions of gRPC and Performance

### ISSUE THREAD

**Issue by juanpedromoreno** [higherkindness/mu-scala#870]
The current WIP in master, after the latest mu-scala 0.21.3 release (using gRPC 1.24.2), is showing a lower performance, apparently because of the usage of a newer version of gRPC and underlying netty version.

I've focused this benchmarks using just [netty](https://netty.io/) as the transport layer.

### mu-scala and gRPC 1.28.x

* Current WIP in the master branch performance (using gRPC 1.28.x) compared to the latest released version.
* Where **Next** <=> Mu using gRPC 1.28.1
👇  
```
| Benchmark                                   |    0.21.3   |        Next |
|---------------------------------------------|-------------|-------------|
| AvroBenchmark.createPerson                  |    1780.620 |    1239.928 |
| AvroBenchmark.getPerson                     |    3099.682 |    2008.010 |
| AvroBenchmark.getPersonLinks                |    1590.598 |    1235.771 |
| AvroBenchmark.listPersons                   |    2344.413 |    1615.139 |
| AvroBenchmark.programComposition            |     281.724 |     268.360 |
| AvroWithSchemaBenchmark.createPerson        |    6768.702 |    2882.209 |
| AvroWithSchemaBenchmark.getPerson           |    8651.390 |    3311.000 |
| AvroWithSchemaBenchmark.getPersonLinks      |    5514.038 |    2703.351 |
| AvroWithSchemaBenchmark.listPersons         |    6858.227 |    2919.710 |
| AvroWithSchemaBenchmark.programComposition  |     978.355 |     844.075 |
| ProtoBenchmark.createPerson                 |   14952.062 |    4140.326 |
| ProtoBenchmark.getPerson                    |   18514.890 |    3966.628 |
| ProtoBenchmark.getPersonLinks               |   11184.247 |    3661.771 |
| ProtoBenchmark.listPersons                  |   12974.525 |    3663.983 |
| ProtoBenchmark.programComposition           |    2726.309 |    1885.634 |
```

### mu-scala and gRPC 1.24.2

* Current WIP in the master branch performance (using gRPC 1.24.2) compared to the latest released version.
* Where **Next** <=> Mu using gRPC 1.24.2
👇  

```
| Benchmark                                   |    0.21.3   |        Next |
|---------------------------------------------|-------------|-------------|
| AvroBenchmark.createPerson                  |    1540.929 |    1679.283 |
| AvroBenchmark.getPerson                     |    2691.254 |    2475.157 |
| AvroBenchmark.getPersonLinks                |    1352.328 |    1268.556 |
| AvroBenchmark.listPersons                   |    2741.178 |    2359.530 |
| AvroBenchmark.programComposition            |     321.900 |     314.986 |
| AvroWithSchemaBenchmark.createPerson        |    7327.412 |    7317.869 |
| AvroWithSchemaBenchmark.getPerson           |    9019.046 |    8533.038 |
| AvroWithSchemaBenchmark.getPersonLinks      |    5196.199 |    5316.770 |
| AvroWithSchemaBenchmark.listPersons         |    7201.592 |    7097.242 |
| AvroWithSchemaBenchmark.programComposition  |    1015.487 |    1007.500 |
| ProtoBenchmark.createPerson                 |   16371.194 |   15676.436 |
| ProtoBenchmark.getPerson                    |   17547.022 |   17637.766 |
| ProtoBenchmark.getPersonLinks               |   10826.200 |   10996.639 |
| ProtoBenchmark.listPersons                  |   12481.453 |   12783.463 |
| ProtoBenchmark.programComposition           |    3021.108 |    3259.551 |
```

### Conclusions

If I'm not missing anything, the reason why mu-scala is providing a worse performance is clearly because of the gRPC version.

### mu-scala and 1.27.2

* **Next** <=> Mu using gRPC 1.27.2
👇 
```
| Benchmark                                   |    0.21.3   |        Next |
|---------------------------------------------|-------------|-------------|
| AvroBenchmark.createPerson                  |    1638.435 |    1036.766 |
| AvroBenchmark.getPerson                     |    2632.940 |    1662.670 |
| AvroBenchmark.getPersonLinks                |    1216.771 |    1044.421 |
| AvroBenchmark.listPersons                   |    2230.070 |    1399.735 |
| AvroBenchmark.programComposition            |     267.606 |     248.627 |
| AvroWithSchemaBenchmark.createPerson        |    6466.185 |    2966.577 |
| AvroWithSchemaBenchmark.getPerson           |    8860.611 |    3266.736 |
| AvroWithSchemaBenchmark.getPersonLinks      |    5364.523 |    2519.063 |
| AvroWithSchemaBenchmark.listPersons         |    6572.509 |    2672.249 |
| AvroWithSchemaBenchmark.programComposition  |     974.005 |     850.050 |
| ProtoBenchmark.createPerson                 |   15895.887 |    3941.838 |
| ProtoBenchmark.getPerson                    |   18101.068 |    4327.778 |
| ProtoBenchmark.getPersonLinks               |   10406.542 |    3353.672 |
| ProtoBenchmark.listPersons                  |   12084.273 |    3688.467 |
| ProtoBenchmark.programComposition           |    2700.435 |    1853.069 |
```

### mu-scala and 1.26.1

* **Next** <=> Mu using gRPC 1.26.1
👇 
```
| Benchmark                                   |    0.21.3   |        Next |
|---------------------------------------------|-------------|-------------|
| AvroBenchmark.createPerson                  |    1142.057 |    1068.473 |
| AvroBenchmark.getPerson                     |    2035.920 |    1908.101 |
| AvroBenchmark.getPersonLinks                |    1052.595 |     965.651 |
| AvroBenchmark.listPersons                   |    1916.344 |    1764.138 |
| AvroBenchmark.programComposition            |     241.964 |     227.391 |
| AvroWithSchemaBenchmark.createPerson        |    5321.475 |    5158.408 |
| AvroWithSchemaBenchmark.getPerson           |    8842.853 |    6406.713 |
| AvroWithSchemaBenchmark.getPersonLinks      |    5242.810 |    3904.302 |
| AvroWithSchemaBenchmark.listPersons         |    7058.938 |    5193.020 |
| AvroWithSchemaBenchmark.programComposition  |    1012.665 |     739.563 |
| ProtoBenchmark.createPerson                 |   12983.579 |   11701.910 |
| ProtoBenchmark.getPerson                    |   13085.494 |   12688.804 |
| ProtoBenchmark.getPersonLinks               |    8322.383 |    8165.060 |
| ProtoBenchmark.listPersons                  |    9461.153 |    8449.480 |
| ProtoBenchmark.programComposition           |    2057.679 |    2079.205 |
```

### mu-scala and 1.25.0

* **Next** <=> Mu using gRPC 1.25.0
👇 
```
| Benchmark                                   |    0.21.3   |        Next |
|---------------------------------------------|-------------|-------------|
| AvroBenchmark.createPerson                  |    1504.775 |    1724.736 |
| AvroBenchmark.getPerson                     |    2506.650 |    2663.970 |
| AvroBenchmark.getPersonLinks                |    1315.542 |    1331.707 |
| AvroBenchmark.listPersons                   |    2336.259 |    2453.789 |
| AvroBenchmark.programComposition            |     311.846 |     294.839 |
| AvroWithSchemaBenchmark.createPerson        |    6857.472 |    6717.698 |
| AvroWithSchemaBenchmark.getPerson           |    7637.273 |    7889.151 |
| AvroWithSchemaBenchmark.getPersonLinks      |    5070.613 |    5167.288 |
| AvroWithSchemaBenchmark.listPersons         |    6769.844 |    6833.602 |
| AvroWithSchemaBenchmark.programComposition  |    1009.687 |    1008.313 |
| ProtoBenchmark.createPerson                 |   15994.439 |   14906.918 |
| ProtoBenchmark.getPerson                    |   15852.595 |   17097.698 |
| ProtoBenchmark.getPersonLinks               |    9820.535 |   10362.682 |
| ProtoBenchmark.listPersons                  |   11493.851 |   12517.153 |
| ProtoBenchmark.programComposition           |    2998.013 |    2980.885 |
```

### Conclusions

In conclusion, I'm interpreting these numbers as a performance degradation of gRPC. Am I missing something?
